File: src/main/scala/com/keithalcock/torchscript/RunTorchScript.scala
package com.keithalcock.torchscript

import com.keithalcock.torchscript.utils.Timer
import org.pytorch.IValue
import org.pytorch.Module
import org.pytorch.Tensor

import java.io.PrintWriter

object RunTorchScript extends App {
  val datapath = "../data/ner-conll/"
  val embedpath = "../data/glove.840B.300d.10f.txt"
  val modelpath = "../data/model.pt"

  val datamodule = new NerDatamodule(datapath, embedpath)
  println(s"Size of dataset: ${datamodule.len}")
  println(s"Size of vocab: ${datamodule.vocabSize}")

  def makeModel(): Module = Module.load(modelpath)

  val model: Module = makeModel()

  def cropInput(tensorPair: (Tensor, Tensor)): Tensor = tensorPair._1

  val timer = new Timer("The timer")

  def runDataset(name: String, dataset: NerDataset, printWriter: PrintWriter): Unit = {
    val times = dataset.map { tensorPair =>
      val cropped = cropInput(tensorPair)
      val result = timer.time {
        model.forward(IValue.from(cropped))
      }
      printWriter.println(result.toTensor.shape.mkString(", "))
      printWriter.println(result.toTensor.getDataAsFloatArray.map(value => f"$value%1.8f").mkString(", ")) // can't be double
      timer.getElapsed
    }
    println(f"  Mean $name sample time: ${timer.mean(times)}%.8f")
    println(f"Stddev $name sample time: ${timer.stddev(times)}%.8f")
  }

  1.to(8).foreach { index =>
    val printWriter = new Sourcenew PrintWriter(System.out)

    runDataset("train", datamodule.trainDataset, printWriter)
    runDataset("  val", datamodule.valDataset, printWriter)
    runDataset(" test", datamodule.testDataset, printWriter)
  }

  System.exit(0);
}

File: src/main/python/run_torchscript.py
import os.path

import torch
import torch.onnx

from NerDatamodule import NerDatamodule
from ScriptedModel import ScriptedModel

from Shared import make_model, run_dataset

def main():
    datapath = "../../../../data/ner-conll/"
    embedpath = "../../../../data/glove.840B.300d.10f.txt"
    modelpath = "../../../../data/model.pt"
    datamodule = NerDatamodule(datapath, embedpath, batch_size = 1)
    example_crop = 50
    crop = 0

    print(f"Size of dataset: {len(datamodule)}")
    print(f"Size of vocab: {len(datamodule.train_dataset.vocab)}")

    def save_model(model, input):
        model.save(modelpath)
        print("The torchscript model was saved.")

    def make_scripted_model(size_of_vocab, size_of_labels):
        if os.path.isfile(modelpath):
            # If the file already exists, just load it.
            model = torch.jit.load(modelpath)
        else:
            example_input = torch.randint(1, size_of_vocab - 1, (1, example_crop))
            model = make_model(size_of_vocab, size_of_labels)
            traced_model = torch.jit.trace(model, example_input)
            save_model(traced_model, example_input)
        return ScriptedModel(model)

    model = make_scripted_model(len(datamodule.train_dataset.vocab), datamodule.num_classes)

    with torch.no_grad():
        run_dataset(model, "train", crop, datamodule.train_dataloader())
        run_dataset(model, "  val", crop, datamodule.val_dataloader())
        run_dataset(model, " test", crop, datamodule.test_dataloader())


if __name__ == "__main__":
    print(torch.__version__)   
    main()
